// inspectable/lib/index.mjs
import { inspect } from 'util';

const inspectable = (klass, { serialize = () => ({}), stringify = (instance, payload, context) => (`${context.stylize(klass.name, 'special')} ${context.inspect(payload)}`) } = {}) => {
    Object.defineProperty(klass.prototype, inspect.custom, {
        value(depth, inspectContext) {
            const context = {
                stylize: inspectContext.stylize,
                inspect: (payload, options) => {
                    var _a;
                    return (inspect(payload, {
                        ...inspectContext,
                        compact: (_a = options === null || options === void 0 ? void 0 : options.compact) !== null && _a !== void 0 ? _a : false
                    }));
                }
            };
            const payload = serialize(this);
            return stringify(this, payload, context);
        }
    });
};

const kInspectProperties = Symbol('kInspectProperties');
const Inspectable = (options = {}) => (
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(klass) => {
    inspectable(klass, {
        ...options,
        serialize(instance) {
            var _a;
            const payload = (((_a = options.serialize) === null || _a === void 0 ? void 0 : _a.call(options, instance)) || {});
            for (const property of (klass[kInspectProperties] || [])) {
                payload[property] = instance[property];
            }
            return payload;
        }
    });
    return klass;
});
const Inspect = (target, property) => {
    if (target.constructor[kInspectProperties] === undefined) {
        target.constructor[kInspectProperties] = [];
    }
    target.constructor[kInspectProperties].push(property);
};

export { Inspect, Inspectable, inspectable, kInspectProperties };
